features/net/network-socket/nsapi_types.h

```diff
@@ -148,6 +148,18 @@ typedef enum nsapi_option {
 } nsapi_option_t;
 
 
+/*  C linkage definition
+ *
+ *  Attribute garuntees emitted function has C linkage.
+ *  Available in both C and C++
+ */
+#ifdef __cplusplus
+#define NSAPI_C_LINKAGE extern "C"
```

Comment:
The `extern "C"` is invalid in actual C code, this macro expands to extern "C" only in C++. The nsapi_dns.h can be included in both C/C++ code, thus the `extern "C"` and switches on `__cplusplus`.